### DISCUSSION TITLE
Rename file on upload

### DISCUSSION THREAD

**Discussion by mdmontesinos** [oqtane/oqtane.framework#2388] [Q&A]
Is it possible to rename a file upon uploading it by using the default FileManager?

For example, getting rid of the original name and asigning a GUID or just a custom standarized name. 

Currently, I've created a method that acts as the OnUpload EventCallback for the FileManager that receives the imageId, changes the name and updates the file using the IFileService.
```c#
private async void OnImageUploaded(int imageId)
    {
        Oqtane.Models.File image = await FileService.GetFileAsync(imageId);
        image.Name = "NEW-NAME";
        await FileService.UpdateFileAsync(image);
    }
```
However, doing this leaves the FileManager in an "inconsistent" state, where the image is set, but not selected automatically.

![image](https://user-images.githubusercontent.com/90258222/187658129-0586f4da-23f3-4f70-8486-efc35f407ada.png)

In addition, if we inspect the file selector, the image appears there with its original name, not the updated one, meaning that the FileManager has not properly updated the list.
![image](https://user-images.githubusercontent.com/90258222/187658487-df6aa24b-a170-421d-82d2-8b13422009e8.png)

If we use _fileManager.GetFile() without selecting any other file (meaning it does show 'Select File'), we get the original updated file with the incorrect name, that does not exist anymore, as the file is succesfully updated in the DB.

So, is there any way to force the FileManager to re-update the list of files in order for this to work properly? Perhaps, using the AfterRenderAsync with some conditional parameters?

EDIT: There, obviously, are workarounds for this problem, but I'd be nice if the FileManager could handle this situation by itself.

**Comment by sbwalker** [comment]
The problem is that Blazor components are independent and stateful. When you hook the OnUpload event in your own component and change the filename, the FileManager component still has the original list of files that it loaded and has no idea that anything changed. There would need to be a new method added to FileManager called Refresh() which would instruct it to reload the list of files and re-render the component UI (by calling StateHasChanged).

**Comment by sbwalker** [comment]
@mdmontesinos #2395 adds the method you need to refresh the list of files. So in your calling component you would need to add a reference and then use that reference to call the Refresh method:

```
<FileManager @ref="filemanager" OnUpload="OnUpload" />
```

```
@code {
    private FileManager filemanager;
````

```
	private async void OnUpload(int fileId)
	{
		Oqtane.Models.File image = await FileService.GetFileAsync(fileId);
		image.Name = "NEW-NAME";
		await FileService.UpdateFileAsync(image);
		await filemanager.Refresh(fileId);
	}
```

Note that there are 2 overloads for Refresh. By default it reloads the list of files and re-renders the component. However if you specify a FileId it will also select it in the list (and display the thumbnail if it is an image).
